### ISSUE TITLE
Dynamic adjustment of pressure advance

### ISSUE THREAD

**Issue by dewi-ny-je** [QIDITECH/QIDI_Q1_Pro#69]
Klipper recently implemented a correction which allows changing pressure advance on the fly without blobs or print defects:
https://github.com/Klipper3d/klipper/pull/6635

Thanks to it, a feature of Orca can now be used:
https://github.com/SoftFever/OrcaSlicer/pull/5609

Let me show you some examples of what the feature allows:

WITH DYNAMIC PA:

Top surface

![337601814-fa0a4155-3546-411d-8023-28464a1dbaed](https://github.com/user-attachments/assets/8efef22d-ba7b-45cc-a6fe-9ef2b9b1ae05)

Solid infill

![337616471-9d949a67-c8b3-477e-9f06-c429d4e40be0](https://github.com/user-attachments/assets/cf107469-f7a5-4a55-a59f-b6b1246baa44)

WITHOUT DYNAMIC PA (pressure advance calibrated, as usual, for corners only):

![337608941-37b1528b-7eab-4721-8dfe-ef4e85840b7a](https://github.com/user-attachments/assets/fe550c80-9709-4326-8845-b82d74e411cd)

The improvements are huge and I think it's worth to backport the Klipper commit linked above.

**Comment by CChen616** [comment]
We will continue to monitor this feature, and if it proves to be stable and without issues, it will be included in a future update.

**Comment by dewi-ny-je** [comment]
> We will continue to monitor this feature, and if it proves to be stable and without issues, it will be included in a future update.

I plan to apply the patch manually in the next days. I'll report back after performing at least 10 prints with it enabled, maybe this will help you evaluating it in real world scenario.

**Comment by dewi-ny-je** [comment]
Right, I forgot it also requires flashing.
So basically this new feature will never come to existing Q1 printers? that's a pity.

I'll do as suggested.

Am I right that, once new firmwares for the Q1 are released by QIDI, I'll just need to manually patch extruder.py without recompiling/flashing the main board and the extruder board?

**Comment by CChen616** [comment]
> Am I right that, once new firmwares for the Q1 are released by QIDI, I'll just need to manually patch extruder.py without recompiling/flashing the main board and the extruder board?

Yes, and to be safe, it's best to back up the `extruder.py` file before updating, as we might modify this file in the new patch.

**Comment by dewi-ny-je** [comment]
@CChen616 
> For extruder board: 
> Press and hold the `boot` button on the extruder board and start the machine. The extruder board should mount as a USB device, and the device directory should be mounted to `gcode_files/sda1/` (assuming there are no other USB devices). Move the generated `klipper.uf2` file into this directory and then reboot.

To be clear: the Q1 extruder board seems connected via serial, not USB, see printer.cfg:

```
[mcu]
serial: /dev/ttyS2
restart_method: command

[mcu U_1]
serial: /dev/ttyS0
restart_method: command
```

Will it really appear as USB mass storage, since it's apparently not USB to begin with?

I was planning to do it now but I was pointed out to this inconsistency.

I later found also https://github.com/frap129/OpenQ1/blob/main/docs/Installation.md which instructs (step 3) to solder new wires to the toolhead for the purpose.

Thanks for the help

**Comment by dewi-ny-je** [comment]
I was right, see https://github.com/QIDITECH/QIDI_Q1_Pro/issues/78#issuecomment-2307991617

**Comment by dewi-ny-je** [comment]
I patched the firmware of my Q1 v4.4.24 and it works.
As you can see from my previous post and the wiki page the results are very good and I think worth a firmware release because the print quality is quite impressive.

https://github.com/SoftFever/OrcaSlicer/wiki/adaptive-pressure-advance

This is the patch without the compiled .pyc files, only the sources I modified.
[dynamic-PA.zip](https://github.com/user-attachments/files/18247077/dynamic-PA.zip)
